### ISSUE TITLE
Confusing documentation/behavior of max_requests and block_size parameters (SFTPClient.put)

### ISSUE THREAD

**Issue by pocek** [ronf/asyncssh#710]
asyncssh v2.18.0 (also v2.17.0).

Hello Ron. I have stumbled upon a confusing SFTPClient.put() behavior while debugging some performance issue. For easy reference, takeout from https://asyncssh.readthedocs.io/en/latest/api.html#asyncssh.SFTPClient.put:

> The block_size argument specifies the size of read and write requests issued when downloading the files, defaulting to the maximum allowed by the server, or 16 KB if the server doesn’t advertise limits.

Side note: shouldn't this read *when uploading the files*?

> The max_requests argument specifies the maximum number of parallel read or write requests issued, defaulting to 128.

With that knowledge, let's get to the point. If one does:

```python
await sftp.put('test.file', '/tmp/test.file', max_requests=1, block_size=32 * 1024)
```

asyncssh does this:
```
asyncssh.sftp|[conn=0, chan=0] Sending write for 16384 bytes at offset 163840 in handle 00000000
asyncssh.sftp|[conn=0, chan=0] Sending write for 16384 bytes at offset 180224 in handle 00000000
asyncssh.sftp|[conn=0, chan=0] Received OK
asyncssh.sftp|[conn=0, chan=0] Received OK
[repeat]
```

So, two write requests run in parallel. Seems weird. This is not what the put() call has requested. If we patch asyncssh.sftp a little bit:

```diff
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index aa3fad2..d2752c6 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -798,7 +798,7 @@ class _SFTPFileCopier(_SFTPParallelIO[int]):
 
         try:
             self._src = await self._srcfs.open(self._srcpath, 'rb')
-            self._dst = await self._dstfs.open(self._dstpath, 'wb')
+            self._dst = await self._dstfs.open(self._dstpath, 'wb', block_size=self._block_size, max_requests=self._max_requests)
 
             if self._progress_handler and self._total_bytes == 0:
                 self._progress_handler(self._srcpath, self._dstpath, 0, 0)
```

then asyncssh does this:
```
asyncssh.sftp|[conn=0, chan=0] Sending write for 32768 bytes at offset 491520 in handle 00000000
asyncssh.sftp|[conn=0, chan=0] Received OK
[repeat]
```
So I would risk saying *as expected*.

Of course this simplistic patch is broken, because it breaks if we reverse the direction (do a get instead of a put), but it's enough to illustrate the problem.

I don't know if this is a bug, but it certainly seems like a weird behavior if I understand things correctly.

**Comment by ronf** [comment]
Thanks for the report!

On your first point, you're absolutely right. That was a copy/paste error from the docs for "get" that didn't get updated along with the other references to upload/uploading.

Does your SFTP server support the "limits" extension? If not, I'm thinking this behavior might be tied to AsyncSSH forcing writes to be no more than the "safe" write size of 16 KB when the server doesn't report back what it supports. That said, it does seem wrong that your block_size request of 32 KB didn't override that. It looks like it did the right thing at the SFTPFileCopier level, but it may be breaking that 32 KB write() down into multiple 16 KB writes because of the smaller block size on the file object (self._dst) that the copier opens. That's consistent with it working better when you passed the block size in on the open. This is tricky, though, as the code is shared by the different get/put/copy/mget/mput/mcopy functions, as not all of the file classes opened there support passing in a block size like that.

Turning off parallel reads/write on the source and destination files opened in SFTPFileCopier would probably be sufficient here. This can be done by setting the block_size to 0 when doing the open. Copying over block_size and max_requests is probably not a good idea, as then you potentially end up with two levels of parallelism.

To allow the block size to be set, I'll probably have to add that as an argument to the open() method in _SFTPFSProtocol, as well as LocalFS (where it would just be ignored). I'll experiment with this and let you know when I have a fix.

**Comment by pocek** [comment]
> Does your SFTP server support the "limits" extension?

No, the server I've been testing with does not support limits:

```
asyncssh|[conn=0] Received version SSH-2.0-OpenSSH_8.4p1 Debian-5+deb11u3
(…)
asyncssh.sftp|[conn=0, chan=0] Starting SFTP client
asyncssh.sftp|[conn=0, chan=0] Sending init, version=3
asyncssh.sftp|[conn=0, chan=0] Received version=3, extensions:
asyncssh.sftp|[conn=0, chan=0]   posix-rename@openssh.com: 1
asyncssh.sftp|[conn=0, chan=0]   statvfs@openssh.com: 2
asyncssh.sftp|[conn=0, chan=0]   fstatvfs@openssh.com: 2
asyncssh.sftp|[conn=0, chan=0]   hardlink@openssh.com: 1
asyncssh.sftp|[conn=0, chan=0]   fsync@openssh.com: 1
asyncssh.sftp|[conn=0, chan=0]   lsetstat@openssh.com: 1
asyncssh.sftp|[conn=0, chan=0] Adjusting for non-standard symlink implementation
```

> Turning off parallel reads/write on the source and destination files opened in SFTPFileCopier would probably be sufficient here. This can be done by setting the block_size to 0 when doing the open.

I know this might not be what you mean here, but it is perhaps worth noting that passing block_size=0, max_requests=1 to put() results in asyncssh happily doing 0 bytes writes:

```
asyncssh.sftp|[conn=0, chan=0] Sending write for 0 bytes at offset 0 in handle 00000000
asyncssh.sftp|[conn=0, chan=0] Received OK
[repeat]
```

:)

**Comment by ronf** [comment]
No - I meant passing in `block_size=0` on the call to 'open()' inside SFTPFileCopier. However, I'll look into what happens with a zero block size on a `put()`.

I notice you've mentioned setting `max_requests = 1`. Are you trying to disable parallel writes there? Doing so will _significantly_ performance, particularly for large files (megabytes).

Regarding the limits extension, it looks like the version you have just barely didn't include it. The server is OpenSSH 8.4 and the limits extension was added in 8.5.

**Comment by pocek** [comment]
> I notice you've mentioned setting max_requests = 1. Are you trying to disable parallel writes there? Doing so will significantly performance, particularly for large files (megabytes).

At first using `max_requests=1` was just a guess while debugging some unusually slow uploads and this led me to creating this issue (I still haven't found the root cause of my issue, but it's irrelevant for **this** issue).  I've used `max_requests=1` above to show the writes clearly for the simple case.

**Comment by ronf** [comment]
Ok - I have a potential fix for these issues in commit ca15e11 in the develop branch. It corrects the doc error you found and a similar one in copy(), as well as the issues with block_size not be honored when the default block size is smaller and the problem you found with passing block_size of 0 to get/put/copy. Thanks so much for reporting these!

Regarding setting max_requests=1, that will generally make SFTP take _much_ longer than if you set it to a higher value. I understand using it here to try and make the problem clearer, but if you're trying to debug performance issues, setting that is probably not a good idea.

Let me know if these fixes work for you...

**Comment by ronf** [comment]
This fix is now available in AsyncSSH 2.19.0.
